ide/editor.completion/src/org/netbeans/spi/editor/completion/ItemFactory.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.spi.editor.completion;
+
+/**
+ * Factory for completion items. Used in {@link CompletionProvider#getCompletionItems(javax.swing.text.Document, int, org.netbeans.spi.editor.completion.ItemFactory)}.
+ *
+ * @author Dusan Balek
+ * @since 1.57
+ */
+public interface ItemFactory<T> {
+
+    /**
+     * Creates completion item.
+     *
+     * @param label
+     * @param kind
+     * @param tags
+     * @param sortText
+     * @param insertText
+     * @param insertTextFormat
+     * @param documentation
+     * @return created completion item
+     */
+    T create(String label, int kind, int[] tags, String sortText, String insertText, int insertTextFormat, String documentation);
```

Comment:
Arguments of this method form both [API as well as SPI](http://wiki.apidesign.org/wiki/APIvsSPI) and it is clear the current set of arguments isn't final, it will need to grow. The usual advice when dealing with both caller&provider API is to _separate them_. Let's do the following:
```java
public interface ItemFactory<T> {
  T create(Params p);
}

public final class Params {
   private Params(...) {...}
   String getLabel();
   String getSortText();
}
```
The `Params` class (name subject to change) represents the client API - e.g. people can only call the getters. Should there be a need for _lazy_ getter, let's return something like `CompletionStage<String>` to defer the start of computation until it is needed (ccing @sdedic).

There are (going to be) many providers constructing the `Params` class. The best way to make it ready for evolution is to use the builder pattern. E.g. YAML, Java, XML, etc. should be able to call something like:
```java
Params.newBuilder().sortText(...).label(....).....build();
```
to get the `Params` instance and send it to the `create` method. With such setup it is possible to expand the number of parameters in the future (add new getter to `Params` and new method to the builder) as well as deprecate some of the getters and replace them with other concept (for example compute `Params.isDeprecated` from a value of `tags`, etc.). Overall it should be possible to evolve the client part as well as provider part without compromising existing usages.

Reply:
API separated from SPI as suggested. Builder pattern used.